### PR TITLE
Add siege weapon actor type

### PR DESCRIPTION
**Pull request by surged20**
Adds a siege weapon actor and a sheet modeled off of the vehicle sheet style.
Icon is derived from https://game-icons.net/1x1/heavenly-dog/catapult.html

### CODE REVIEW

**Comment by stwlam** [comment]
This isn't something we can consider for this release, and next major release will be for V10--which will require a full rewrite of the PR.

**Comment by surged20** [comment]
> This isn't something we can consider for this release, and next major release will be for V10--which will require a full rewrite of the PR.

I figured that might be the case for this type of enhancement given the focus on v10. I'll plan to update the PR to the v10 release when available.

**Comment by stwlam** [comment]
OK, sounds good. Thank you for the contribution, though, and sorry about the timing!

**Comment by surged20** [comment]
I'm resolving a UI issue and will fix the eslint fallout along with that.

**Comment by surged20** [comment]
This should pass CI now if you can approve the workflow.

File: packs/scripts/run-migration.ts [hunk 1]
```diff
@@ -76,7 +76,7 @@ const packsDataPath = path.resolve(process.cwd(), "packs/data");
 
 type CompendiumSource = CompendiumDocument["_source"];
 
-const actorTypes = ["character", "npc", "hazard", "loot", "familiar", "vehicle"];
+const actorTypes = ["character", "npc", "hazard", "loot", "familiar", "siegeweapon", "vehicle"];
```
Comment (stwlam) [inline packs/scripts/run-migration.ts:79]: camel case

File: src/module/actor/siegeweapon/data.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    BaseHitPointsData,
+    BaseTraitsData,
+} from "@actor/data/base";
+import { SaveType } from "@actor/types";
+import { ActorSizePF2e } from "@actor/data/size";
+import { SiegeWeaponPF2e } from ".";
+import { SiegeWeaponTrait } from "./types";
+
+/** The stored source data of a siege weapon actor */
+type SiegeWeaponSource = BaseActorSourcePF2e<"siegeweapon", SiegeWeaponSystemData>;
```
Comment (stwlam) [inline src/module/actor/siegeweapon/data.ts:16]: Separate source from data, even if there's no difference except an alias

Comment (surged20) [reply src/module/actor/siegeweapon/data.ts:16]: ok, something more like this?
```
/** The stored source data of a siege weapon actor */
type SiegeWeaponSource = BaseActorSourcePF2e<"siegeWeapon", SiegeWeaponSystemSource>;

interface SiegeWeaponData
    extends Omit<SiegeWeaponSource, "effects" | "flags" | "items" | "prototypeToken">,
        BaseActorDataPF2e<SiegeWeaponPF2e, "siegeWeapon", SiegeWeaponSystemData, SiegeWeaponSource> {}

/** The system-level data of siege weapon actors. */
interface SiegeWeaponSystemSource extends ActorSystemSource {
    attributes: SiegeWeaponAttributesSource;
    details: SiegeWeaponDetailsSource;
    saves: SiegeWeaponSavesSource;
    traits: SiegeWeaponTraitsSource;
}

interface SiegeWeaponSystemData extends SiegeWeaponSystemSource {
    attributes: SiegeWeaponAttributesData;
    details: SiegeWeaponDetailsData;
}
```

Comment (stwlam) [reply src/module/actor/siegeweapon/data.ts:16]: Yeah, that's it.

File: src/module/actor/siege-weapon/data.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        long: number;
+        wide: number;
+        high: number;
+    };
+    crew: {
+        min: number;
+        max: number;
+    };
+    proficiency: string;
+    AC: number;
```
Comment (stwlam) [inline src/module/actor/siege-weapon/data.ts:80]: Why AC in both details and attributes?

File: src/module/actor/siege-weapon/data.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    speed: number;
+    propulsion: string;
```
Comment (stwlam) [inline src/module/actor/siege-weapon/data.ts:82]: Let's pull this from an enumerated list of valid propulsion methods.

**Review by stwlam — CHANGES_REQUESTED** [review]

File: src/module/actor/siege-weapon/data.ts (hunk 3 above)
Comment (surged20) [reply src/module/actor/siege-weapon/data.ts:80]: Good catch, that's some old copy/paste that I missed. Removed on this next update.

File: src/module/actor/siege-weapon/data.ts (hunk 4 above)
Comment (surged20) [reply src/module/actor/siege-weapon/data.ts:82]: Will do

File: src/module/scene/token-document/document.ts [hunk 5]
```diff
@@ -295,8 +295,8 @@ class TokenDocumentPF2e<TActor extends ActorPF2e = ActorPF2e> extends TokenDocum
             huge: 3,
             grg: Math.max(token.width, 4),
         }[actor.size];
-        if (actor.isOfType("vehicle")) {
-            // Vehicles can have unequal dimensions
+        if (actor.isOfType("siegeWeapon") || actor.isOfType("vehicle")) {
```
Comment (stwlam) [inline src/module/scene/token-document/document.ts:298]:
```suggestion
        if (actor.isOfType("siegeWeapon", vehicle")) {
```

**Review by stwlam — CHANGES_REQUESTED** [review]
I think this is close to mergable. Resolve conflicts (sorry!), and rename all *.html files to *hbs.

File: src/module/scene/token-document/document.ts (hunk 5 above)
Comment (surged20) [reply src/module/scene/token-document/document.ts]: Great, no worries, I already had the hbs change pending locally. Fixed conflicts and incorporated your suggested clean up as well.

**Comment by surged20** [comment]
The only concern I have about merging this is that I'm now not happy with the data model of storing mounted siege weapon stats as actions. I started working on converting this to have a fixed aim, load, launch set of data which I think is better for future system automation. e.g. launch would have traits, damage, target type, and save DC properties. In the current model it's just an unstructured wall of text in the Launch action. It's not a huge amount of data entry, but it would all need to change for this data model update.

**Pull request closed without merging**